#Universidad de las Américas
#Progreso II
Edwing García
Inteligencia Artificial I


---



PROGRESO II – Inteligencia Artificial I
Examen de Aprendizaje Supervisado: Modelos de Clasificación
Instrucciones
Utiliza los archivos hospital_data 1.csv y students_data 1.csv para responder las siguientes
preguntas. Escribe tu código en las celdas correspondientes.
1. Cargar los datos
Carga ambos archivos CSV en dos DataFrames de pandas.


In [77]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
import matplotlib.pyplot as plt
import seaborn as sns


df1=pd.read_csv('hospital_data.csv')
df2=pd.read_csv('students_data.csv')

2. Exploración inicial
Muestra las primeras 5 filas y las estadísticas descriptivas de cada conjunto de datos.


In [36]:
df1.head()

Age  Blood_Pressure  Cholesterol  Heart_Rate  Gender  Diagnosis  \
0  0.063332       -0.547694    -0.036711    1.399138       1          0   
1  1.450401       -1.440598     0.970593   -1.473657       0          1   
2 -1.188414       -0.341639    -0.708247    0.789757       1          2   
3  0.739951        1.135087     0.052827   -1.299548       1          2   
4  0.367810       -0.272954    -0.529170   -0.167841       0          2   

   Smoker  Exercise Risk_Level  
0       0         2  High Risk  
1       0         0  High Risk  
2       0         0   Low Risk  
3       1         2  High Risk  
4       1         0  High Risk

In [41]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Age             1000 non-null   int64 
 1   Blood_Pressure  1000 non-null   int64 
 2   Cholesterol     1000 non-null   int64 
 3   Heart_Rate      1000 non-null   int64 
 4   Gender          1000 non-null   object
 5   Diagnosis       1000 non-null   object
 6   Smoker          1000 non-null   object
 7   Exercise        677 non-null    object
 8   Risk_Level      1000 non-null   object
dtypes: int64(4), object(5)
memory usage: 70.4+ KB


In [34]:
df2.head()

GPA  Attendance  Study_Hours  Projects_Completed  Major  Year  \
0  1.308064   -0.541218     1.557873           -1.637660      1     1   
1  0.060327   -0.261338    -0.548310            1.515729      1     2   
2  0.145789    1.487912    -0.109522            0.814976      2     2   
3 -0.828471   -0.611188     0.855812            1.165353      3     0   
4  0.923487    1.487912    -1.338128            0.464599      1     3   

   Scholarship  Extracurricular Result  
0            1                1   Fail  
1            1                3   Fail  
2            0                3   Pass  
3            0                1   Pass  
4            0                2   Fail

In [33]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   GPA                 1000 non-null   float64
 1   Attendance          1000 non-null   float64
 2   Study_Hours         1000 non-null   float64
 3   Projects_Completed  1000 non-null   float64
 4   Major               1000 non-null   int64  
 5   Year                1000 non-null   int64  
 6   Scholarship         1000 non-null   int64  
 7   Extracurricular     1000 non-null   int64  
 8   Result              1000 non-null   object 
dtypes: float64(4), int64(4), object(1)
memory usage: 70.4+ KB


3. Análisis de la variable objetivo
Muestra la distribución de la variable Risk_Level en hospitaldf y Result en studentsdf.


In [42]:
print("Distribucion del Nivel de Riesgo en el Hospital:")
print(df1['Risk_Level'].value_counts())
print("\nDistribucion del Resultado en los Estudiantes:")
print(df2['Result'].value_counts())

Distribucion del Nivel de Riesgo en el Hospital:
Risk_Level
High Risk    503
Low Risk     497
Name: count, dtype: int64

Distribucion del Resultado en los Estudiantes:
Result
Fail    508
Pass    492
Name: count, dtype: int64


4. Preprocesamiento
• Codifica las variables categóricas.
• Normaliza las variables numéricas si es necesario.


In [70]:
columnas_categoricas = df1.select_dtypes(include=['object']).columns
for columna in columnas_categoricas:
    if columna != 'Risk_Level':
        le = LabelEncoder()
        df1[columna] = le.fit_transform(df1[columna])

columnas_numericas = df1.select_dtypes(include=['int64', 'float64']).columns

if 'Risk_Level' in columnas_numericas:
    columnas_numericas = columnas_numericas.drop('Risk_Level')

scaler_hospital = StandardScaler()

df1[columnas_numericas] = scaler_hospital.fit_transform(df1[columnas_numericas])

columnas_categoricas_estudiantes = df2.select_dtypes(include=['object']).columns
for columna in columnas_categoricas_estudiantes:
    if columna != 'Result':
        le = LabelEncoder()
        df2[columna] = le.fit_transform(df2[columna])

columnas_numericas_estudiantes = df2.select_dtypes(include=['int64', 'float64']).columns

if 'Result' in columnas_numericas_estudiantes:
    columnas_numericas_estudiantes = columnas_numericas_estudiantes.drop('Result')

scaler_estudiantes = StandardScaler()
df2[columnas_numericas_estudiantes] = scaler_estudiantes.fit_transform(df2[columnas_numericas_estudiantes])


print("\nDatosPreprocesados hospital:")
print(df1.head())
print("\nDatosPreprocesados estudiantes:")
print(df2.head())


DatosPreprocesados hospital:
        Age  Blood_Pressure  Cholesterol  Heart_Rate    Gender  Diagnosis  \
0  0.063332       -0.547694    -0.036711    1.399138  0.964625  -1.373957   
1  1.450401       -1.440598     0.970593   -1.473657 -1.036672  -0.477703   
2 -1.188414       -0.341639    -0.708247    0.789757  0.964625   0.418550   
3  0.739951        1.135087     0.052827   -1.299548  0.964625   0.418550   
4  0.367810       -0.272954    -0.529170   -0.167841 -1.036672   0.418550   

     Smoker  Exercise  Risk_Level  
0 -0.990050  1.252559           0  
1 -0.990050 -1.186699           0  
2 -0.990050 -1.186699           1  
3  1.010051  1.252559           0  
4  1.010051 -1.186699           0  

DatosPreprocesados estudiantes:
        GPA  Attendance  Study_Hours  Projects_Completed     Major      Year  \
0  1.308064   -0.541218     1.557873           -1.637660 -0.419702 -0.454886   
1  0.060327   -0.261338    -0.548310            1.515729 -0.419702  0.431831   
2  0.145789    1.4

5. División de datos
Divide los datos en conjuntos de entrenamiento y prueba (80/20).



In [71]:
X_hospital = df1.drop('Risk_Level', axis=1)
y_hospital = df1['Risk_Level']

X_train_hospital, X_test_hospital, y_train_hospital, y_test_hospital = train_test_split(X_hospital, y_hospital, test_size=0.2, random_state=42)

X_students = df2.drop('Result', axis=1)
y_students = df2['Result']

X_train_students, X_test_students, y_train_students, y_test_students = train_test_split(X_students, y_students, test_size=0.2, random_state=42)

print("Division de datos para hospital_data:")
print(f"X_train_hospital shape: {X_train_hospital.shape}")
print(f"X_test_hospital shape: {X_test_hospital.shape}")
print(f"y_train_hospital shape: {y_train_hospital.shape}")
print(f"y_test_hospital shape: {y_test_hospital.shape}")

print("\nDivision de datos para students_data:")
print(f"X_train_students shape: {X_train_students.shape}")
print(f"X_test_students shape: {X_test_students.shape}")
print(f"y_train_students shape: {y_train_students.shape}")
print(f"y_test_students shape: {y_test_students.shape}")

Division de datos para hospital_data:
X_train_hospital shape: (800, 8)
X_test_hospital shape: (200, 8)
y_train_hospital shape: (800,)
y_test_hospital shape: (200,)

Division de datos para students_data:
X_train_students shape: (800, 8)
X_test_students shape: (200, 8)
y_train_students shape: (800,)
y_test_students shape: (200,)


6. Entrenamiento de modelos
Entrena al menos dos modelos de clasificación (por ejemplo, Regresión Logística y Árbol
de Decisión) para cada conjunto de datos.

In [72]:
regresion_logistica = LogisticRegression(random_state=42)
regresion_logistica.fit(X_train_hospital, y_train_hospital)
print("- Regresión Logística: regresion_logistica")



- Regresión Logística: regresion_logistica


In [73]:
arbol_decision_hospital = DecisionTreeClassifier(random_state=42)
arbol_decision_hospital.fit(X_train_hospital, y_train_hospital)
print("- Árbol de Decisión: arbol_decision_hospital")

- Árbol de Decisión: arbol_decision_hospital


In [74]:

svm_estudiantes = SVC(random_state=42)
svm_estudiantes.fit(X_train_students, y_train_students)

knn_estudiantes = KNeighborsClassifier()
knn_estudiantes.fit(X_train_students, y_train_students)

print("SVM: svm_estudiantes")
print("KNN: knn_estudiantes")


Additional models trained for students_data:
- Support Vector Machine (SVM): svm_students
- k-Nearest Neighbors (KNN): knn_students


7. Evaluación de modelos
Evalúa los modelos usando métricas como precisión, recall, F1-score y matriz de
confusión.



In [82]:

y_pred_lr_hospital = regresion_logistica.predict(X_test_hospital)
print("\n Evaluación del modelo de Regresión Logística (Hospital)")
print(f"Precisión: {accuracy_score(y_test_hospital, y_pred_lr_hospital):.4f}")
print(f"Recall: {recall_score(y_test_hospital, y_pred_lr_hospital, average='weighted'):.4f}")
print(f"F1-score: {f1_score(y_test_hospital, y_pred_lr_hospital, average='weighted'):.4f}")
conf_matrix_lr_hospital = confusion_matrix(y_test_hospital, y_pred_lr_hospital)
print("\nMatriz de Confusión:")
print(conf_matrix_lr_hospital)

y_pred_dt_hospital = arbol_decision_hospital.predict(X_test_hospital)
print("\n Evaluación del modelo de Árbol de Decisión (Hospital)")
print(f"Precisión: {accuracy_score(y_test_hospital, y_pred_dt_hospital):.4f}")
print(f"Recall: {recall_score(y_test_hospital, y_pred_dt_hospital, average='weighted'):.4f}")
print(f"F1-score: {f1_score(y_test_hospital, y_pred_dt_hospital, average='weighted'):.4f}")
conf_matrix_dt_hospital = confusion_matrix(y_test_hospital, y_pred_dt_hospital)
print("\nMatriz de Confusión:")
print(conf_matrix_dt_hospital)





 Evaluación del modelo de Regresión Logística (Hospital)
Precisión: 0.4700
Recall: 0.4700
F1-score: 0.4652

Matriz de Confusión:
[[53 38]
 [68 41]]

 Evaluación del modelo de Árbol de Decisión (Hospital)
Precisión: 0.5150
Recall: 0.5150
F1-score: 0.5152

Matriz de Confusión:
[[43 48]
 [49 60]]


In [80]:
y_pred_svm_students = svm_students.predict(X_test_students)
print("\n--- Evaluación del modelo SVM (Estudiantes)")
print(f"Precisión: {accuracy_score(y_test_students, y_pred_svm_students):.4f}")
print(f"Recall: {recall_score(y_test_students, y_pred_svm_students, average='weighted'):.4f}")
print(f"F1-score: {f1_score(y_test_students, y_pred_svm_students, average='weighted'):.4f}")
conf_matrix_svm_students = confusion_matrix(y_test_students, y_pred_svm_students)
print("\nMatriz de Confusión:")
print(conf_matrix_svm_students)



y_pred_knn_students = knn_students.predict(X_test_students)
print("\n--- Evaluación del modelo KNN (Estudiantes)")
print(f"Precisión: {accuracy_score(y_test_students, y_pred_knn_students):.4f}")
print(f"Recall: {recall_score(y_test_students, y_pred_knn_students, average='weighted'):.4f}")
print(f"F1-score: {f1_score(y_test_students, y_pred_knn_students, average='weighted'):.4f}")
conf_matrix_knn_students = confusion_matrix(y_test_students, y_pred_knn_students)
print("\nMatriz de Confusión:")
print(conf_matrix_knn_students)




--- Evaluación del modelo SVM (Estudiantes)
Precisión: 0.4750
Recall: 0.4750
F1-score: 0.4722

Matriz de Confusión:
[[59 48]
 [57 36]]

--- Evaluación del modelo KNN (Estudiantes)
Precisión: 0.4550
Recall: 0.4550
F1-score: 0.4557

Matriz de Confusión:
[[49 58]
 [51 42]]


8. Comparación de modelos
Compara el rendimiento de los modelos y justifica cuál elegirías para cada conjunto de
datos.
-----------------------------------------------------------------------------

Entre los modelos usados para el hospital (regresión logística y arbol de decisión) se puede notar como el árbol de decisión posee una precisión o acurrancy del 0.51 un 4.5% más que el de la regresión logística, por lo que este modelo es el mejor entre esos dos.

-------------------------------------------------------

En cuanto al dataset de los estudiantes se entrenaron dos modelos diferentes(SVM y KNN) donde en este caso el modelo SVM posee un valor de 0.47 en el indicador de precisión con un 2%  por encima del modelo de KNN indicando que para este caso es la mejor opción